draft-ietf-httpbis-connect-tcp.md

```diff
@@ -145,13 +152,12 @@ Authentication to a templated TCP proxy normally uses ordinary HTTP authenticati
 
 Clients SHOULD assume that all proxy resources generated by a single template share a protection space (i.e., a realm) ({{?RFC9110, Section 11.5}}).  For many authentication schemes, this will allow the client to avoid waiting for a "401 (Unauthorized)" response before each new connection through the proxy.
 
-## Use of the Capsule Protocol {#capsule-protocol}
-
-When using the "connect-tcp" Upgrade Token, templated TCP proxies do not use the Capsule Protocol {{!RFC9297}}.  Clients MAY request use of the Capsule Protocol by offering the Upgrade Token "connect-tcp-capsule" instead.  When offering or accepting the "connect-tcp-capsule" Upgrade Token, clients and servers SHOULD include a "Capsule-Protocol: ?1" header to facilitate processing by intermediaries.
+## Closing Connections
 
-Clients of this specification MAY implement "connect-tcp", "connect-tcp-capsule", or both.  Accordingly, a templated TCP proxy server MUST implement both Upgrade Tokens unless its use is restricted to a subset of compatible clients.
+In each HTTP version, any requirements related to closing connections in Classic HTTP CONNECT also apply to "connect-tcp", with the following modifications:
 
-When using "connect-tcp-capsule", TCP payload data is sent in the payload of a new Capsule Type named DATA ({{data-capsule}}).  The ordered concatenation of DATA capsule payloads represents the main payload data stream in any protocol where this is well-defined.  Intermediaries MAY split or merge DATA capsules.  Endpoints MAY indicate a TCP connection error by sending an incomplete DATA capsule, as an alternative to using TCP, TLS, or HTTP stream errors.
+* In HTTP/1.1, endpoints SHOULD send an incomplete DATA capsule to indicate receipt of a TCP connection error (e.g., a TCP RST or timeout).  When a connection is terminated with any incomplete capsule, or with an error in the underlying connection (e.g. a TLS "internal_error" alert or TCP RST), the endpoint SHOULD send a TCP RST if the underlying TCP implementation permits it.
```

Comment:
What's "incomplete DATA capsule"?

Reply:
What if there is no data in-flight (i.e. an idle connection)? Should intermediaries inject malformed `DATA` frame?

Reply:
Since we're now always using capsules, I think there's value in making this an explicit signal. How about adding a DATA_END capsule of length_zero and requiring to send that? If the stream closes without receiving that, then it means it was aborted.

Alternate encoding: is to have two capsule types for DATA and a FIN bit, like we do for STREAM frames in QUIC.

Reply:
BIKESHED TIME

There are a bunch of different ways a TCP connection can end.  At a minimum, there's the distinction between FIN and RST, but there's also timeout (dead air), interface going down, ICMP Unreachable, proxy-injected errors like "quota exceeded", etc.  We have to decide how much detail to encode, and how to encode it.

The simplest option, as you say, would probably be something like an empty DATA_END capsule indicating clean shutdown, with RST assumed otherwise.

Reply:
FIN vs RST is a useful distinction. It's something we accomodate in H2 and H3 CONNECT. I'd be happy to consider those in scope to address but punt on others

Reply:
I don't think we need `DATA_END` capsule to signal a clean shutdown, we could use an empty `DATA` for that. This is how BSD sockets and other applications transmit `FIN` to userland.

But yes, using either empty `DATA` or adding a `flags` field with a `FIN` bit to signal a clean shutdown, and using the lack of that signal and a termination of the underlying transport to signal connection reset seems better than injecting malformed `DATA` frames.

Reply:
> There are a bunch of different ways a TCP connection can end. At a minimum, there's the distinction between FIN and RST, but there's also timeout (dead air), interface going down, ICMP Unreachable, proxy-injected errors like "quota exceeded", etc. We have to decide how much detail to encode, and how to encode it.

We have to be able to emit those back to TCP endpoints, so the amount of details and bike-sheding that we can do here is limited.

Reply:
DID SOMEONE SAY BIKESHED??? I'M ENTHUSIASTIC

I never liked the fact that reading 0 on a BSD socket means "end of stream" for TCP. It's odd, and inconsistent with other uses (e.g., for UDP it means "I received a UDP packet with no payload").

We should definitely separate between "clean termination" (TCP FIN) and "abortive termination" (TCP RST, timeout, etc.) but I don't think we need to separate RST from timeout. That would match what classic CONNECT does.

So, +1 for DATA capsule with a FIN bit :)

Reply:
A FIN bit adds overhead to every DATA capsule, which seems unappealing to me.  A zero-length DATA capsule seems fine to me (AFAIK we don't have any other use for length=0), but otherwise I would define a new capsule type ... probably a type meaning "clean shutdown", so that underlying transport failures naturally are treated as "unclean".

Reply:
Sorry I meant putting the FIN bit in the capsule type. That can be described as two separate capsules (DATA = 0xTBD, DATA_END=0xTBD+1) or as a bit (DATA = 0xTBD..0xTBD+1, lowest bit of capsule type means FIN) but they're the same thing on the wire